### ISSUE TITLE
Gan网络

### ISSUE THREAD

**Issue by MiniBullLab** [MiniBullLab/easy_ai#3]
1. 手写体生成
2. 异常检测

**Comment by MiniBullLab** [comment]
手写体GAN训练代码链接
@kingwangxiang

**Comment by lpj0822** [comment]
https://blog.csdn.net/happyday_d/article/details/84961175

**Comment by MiniBullLab** [comment]
https://blog.csdn.net/qq_40268255/article/details/103289290

**Comment by foww-0001** [comment]
异常检测：https://zhuanlan.zhihu.com/p/44900293

**Comment by lpj0822** [comment]
手写体生成代码编写完成

**Comment by kingwangxiang** [comment]
1.使用easy ai平台测试Gan手写体数字图片生成任务
1) -t 参数传入generate_image，程序报This task(generate_image) not exits!
![minist_gan](https://user-images.githubusercontent.com/37147736/101270586-520b1b80-37b5-11eb-868f-2e32bce4cee1.png)

**Comment by MiniBullLab** [comment]
修改好了，运行
```
python3 easyai/train_task.py -t generate_image -i /home/easyai/easy_data/caltech100_classify/ImageSets/train.txt -v /home/easyai/easy_data/caltech100_classify/ImageSets/val.txt -m MNISTGan
```

**Comment by kingwangxiang** [comment]
1.使用easy ai最新代码测试Gan手写体数字图片生成任务
1)代码bug已修复，可以run，但是训练结果不对，计算出的loss有问题，如图1-A所示。
![gan_loss](https://user-images.githubusercontent.com/37147736/102705294-dcc53d80-42c0-11eb-9331-3e6da6c7f429.png)
2)初步调试代码，输入判别器的数据不对(如图2-A)，计算loss的标签不对如图2-B。
![minist_gan](https://user-images.githubusercontent.com/37147736/102705291-d46d0280-42c0-11eb-8b94-f1f91f98ed14.png)
![test](https://user-images.githubusercontent.com/37147736/102705285-cae39a80-42c0-11eb-8970-a8917b532654.png)

**Comment by kingwangxiang** [comment]
1.使用easy ai最新代码测试Gan手写体数字图片生成任务
1)使用生成器梯度更新1次判别器更新5次的训练策略训练网络。网络训练稳定性稍有改进，训练初始阶段g loss有明显下降，d loss趋于一个值来回波动。到后期，g loss值很难下降，d loss值有所增大。训练出的模型测试结果如下图1-A所示，测试结果并没明显的辨识出是数字，但显示的图片样式，大致符合手写体数字图片分布样式。
![0road_46592](https://user-images.githubusercontent.com/37147736/104115689-d587d180-534c-11eb-8d2f-00c806c41a73.png)
![0road_48035](https://user-images.githubusercontent.com/37147736/104115693-dcaedf80-534c-11eb-80b5-872535879339.png)
2)后续尝试新的学习率，再调整下模型。

**Comment by kingwangxiang** [comment]
调整的训练策略：
    def compute_backward(self, input_datas, targets, setp_index):
        # Compute loss, compute gradient, update parameters
        for i in range(5):
            d_loss_values = self.discriminator_backward(input_datas, targets)
        g_loss_values = self.generator_backward(input_datas, targets)
        return d_loss_values, g_loss_values

**Comment by kingwangxiang** [comment]
以前尝试训练过gan生成的minist手写体数字的代码，其中输出图片做了适当的放大。
···
import  torch
from    torch import nn, optim, autograd #autograd手动对X^求倒
import  numpy as np
import  visdom
import  random
from    matplotlib import pyplot as plt




h_dim = 400
batchsz = 512
viz = visdom.Visdom()

class Generator(nn.Module):

    def __init__(self):
        super(Generator, self).__init__()

        self.net = nn.Sequential(
            # z: [b, 2] => [b, 2]
            nn.Linear(2, h_dim),
            nn.ReLU(True),
            nn.Linear(h_dim, h_dim),
            nn.ReLU(True),
            nn.Linear(h_dim, h_dim),
            nn.ReLU(True),
            nn.Linear(h_dim, 2)
        )

    def forward(self, z):
        output = self.net(z)
        return output


class Discriminator(nn.Module):

    def __init__(self):
        super(Discriminator, self).__init__()

        self.net = nn.Sequential(
            nn.Linear(2, h_dim),
            nn.ReLU(True),
            nn.Linear(h_dim, h_dim),
            nn.ReLU(True),
            nn.Linear(h_dim, h_dim),
            nn.ReLU(True),
            nn.Linear(h_dim, 1),
            nn.Sigmoid()
        )

    def forward(self, x):
        output = self.net(x)
        return output.view(-1)



def data_generator():
    '''
    8-gaussion mixture models
    :return:
    '''
    scale = 2
    centers = [
        (1, 0),
        (-1, 0),
        (0, 1),
        (0, -1),
        (1. / np.sqrt(2), 1. / np.sqrt(2)),
        (1. / np.sqrt(2), -1. / np.sqrt(2)),
        (-1. / np.sqrt(2), 1. / np.sqrt(2)),
        (-1. / np.sqrt(2), -1 / np.sqrt(2))
    ]
    centers = [(scale * x, scale * y) for x, y in centers]

    # 通过 yield 制作一个无限循环数据生成器
    # (这里yield每次循环会暂停一次，下次接着上次遍历的点继续往下迭代，直到完成呢整个数据遍历)
    while True:
        dataset = []

        for i in range(batchsz):
            point = np.random.randn(2) * 0.02
            center = random.choice(centers)
            # N(0,1) + center x1/x2
            point[0] += center[0]
            point[1] += center[1]
            dataset.append(point)

        dataset = np.array(dataset).astype(np.float32)
        dataset /= 1.414
        yield dataset

def generate_image(D, G, xr, epoch):
    """
    Generates and saves a plot of the true distribution, the generator, and the
    critic.
    """
    N_POINTS = 128
    RANGE = 3
    plt.clf()

    points = np.zeros((N_POINTS, N_POINTS, 2), dtype='float32')
    points[:, :, 0] = np.linspace(-RANGE, RANGE, N_POINTS)[:, None]
    points[:, :, 1] = np.linspace(-RANGE, RANGE, N_POINTS)[None, :]
    points = points.reshape((-1, 2))
    # (16384, 2)
    # print('p:', points.shape)

    # draw contour
    with torch.no_grad():
        points = torch.Tensor(points).cuda() # [16384, 2]
        disc_map = D(points).cpu().numpy() # [16384]
    x = y = np.linspace(-RANGE, RANGE, N_POINTS)
    cs = plt.contour(x, y, disc_map.reshape((len(x), len(y))).transpose())
    plt.clabel(cs, inline=1, fontsize=10)
    # plt.colorbar()


    # draw samples
    with torch.no_grad():
        z = torch.randn(batchsz, 2).cuda() # [b, 2]
        samples = G(z).cpu().numpy() # [b, 2]
    plt.scatter(xr[:, 0], xr[:, 1], c='orange', marker='.')
    plt.scatter(samples[:, 0], samples[:, 1], c='green', marker='+')

    viz.matplot(plt, win='contour', opts=dict(title='p(x):%d'%epoch))



def main():

    torch.manual_seed(23)
    np.random.seed(23)

    device = torch.device('cuda:0')

    data_iter = data_generator()
    x = next(data_iter)
    # [b, 2]
    # print(x.shape)

    G = Generator().to(device)
    D = Discriminator().to(device)
    # print(G)
    # print(D)
    lr = 5e-4
    optim_G = optim.Adam(G.parameters(), lr=lr, betas=(0.5, 0.9))
    optim_D = optim.Adam(D.parameters(), lr=lr, betas=(0.5, 0.9))

    viz.line([[0, 0]], [0], win='loss', opts=dict(title='loss', legend=['D', "G"]))

    for epoch in range(50000):
        # 1. train Discriminator firstly
        for _ in range(5):
            # 1.1 train on real data
            xr = next(data_iter)
            xr = torch.from_numpy(xr).to(device)

            # [b, 2] => [b, 1]
            predr = D(xr)
            # max predr, min lossr
            lossr = -predr.mean()

            # 1.2 train on fake data
            # [b, 2]
            z = torch.randn(batchsz, 2).to(device) # 这里的z是隐藏变量，是指什么？？
            # 通过z生成fake data 注意这里只train G,
            # 但是创建D的计算图的时候,G(z)也创建到计算图中了，所以要打断它，不让它更新G网络
            # 计算图返向传播,是从最后出来的结果往回逆运算, 梯度先传到D网络然后再传到G网络，当
            # 到传到G网络的时候，截断不往回更新参数。
            # 注意: 主要看输入输出流经的网络的，先后过程，梯度也会按着这个反顺序,往回一次跟新参数
            xf = G(z).detach()# 类似TensorFlow中tf.stop_gradient()
            predf = D(xf)
            lossf = predf.mean()

            # aggregate all
            loss_D = lossr + lossf

            # optimize
            optim_D.zero_grad()
            loss_D.backward()
            optim_D.step()

        # 2. train Generator
        z = torch.randn(batchsz, 2).to(device)
        xf = G(z)
        predf = D(xf)
        # max predf.mean()
        loss_G = -predf.mean()

        # optimize
        optim_G.zero_grad()
        loss_G.backward()
        optim_G.step()

        if epoch % 100 == 0:
            viz.line([[loss_D.item(), loss_G.item()]], [epoch], win='loss', update='append')


            generate_image(D, G, xr.cpu(), epoch)
            print(loss_D.item(), loss_G.item())

if __name__ == '__main__':
    main()

···

**Comment by kingwangxiang** [comment]
1.使用easy ai最新代码测试Gan生成手写体数字图片任务
1)代码已调通，训练gan生成minist手写体数字图片，可以初步生成图手写体数字。模型迭代中间结果如图1-A所示。
![gan生成手写体数字图片](https://user-images.githubusercontent.com/37147736/104830649-919c4b80-58bb-11eb-91d8-08a5d507da06.png)
1-A模型迭代中间结果

**Comment by kingwangxiang** [comment]
1.尝试跑通杜伦大学提出GANomaly：无需负例样本实现异常检测
1)搭建环境遇到问题，一直卡在安装conda install -c intel mkl_fft。目前还没有想到如何解决。
![ganormal](https://user-images.githubusercontent.com/37147736/106376252-7a3a8380-63ce-11eb-838b-065379967984.PNG)

**Comment by lpj0822** [comment]
https://www.lfd.uci.edu/~gohlke/pythonlibs/#mkl_fft

**Comment by kingwangxiang** [comment]
1.尝试跑通杜伦大学提出GANomaly：无需负例样本实现异常检测
1)依赖环境已配置完成，尝试复现论文中的结果。
2)代码已跑通，模型训练中如图2-A所示。模型迭代的中间结果，如图2-B(假图，生成器生成)、2-C(真图)所示。
![ganomaly_train](https://user-images.githubusercontent.com/37147736/108616143-c153eb80-7445-11eb-80fe-d4cbebb3c14e.PNG)
2-A模型训练中
![fakes](https://user-images.githubusercontent.com/37147736/108616189-1d1e7480-7446-11eb-8ed0-f4e6d8bad93c.png)
2-B 模型生成的图片
![reals](https://user-images.githubusercontent.com/37147736/108616194-2a3b6380-7446-11eb-9a77-b23090ed643a.png)
2-C真图
3)当前代码跑出的一些结果还不太清楚对应些什么，还要对照论文细致研究下。

**Comment by lpj0822** [comment]
看一下能不能区分异常图片，也就是对于输入一张图片，最后输出是否异常

**Comment by kingwangxiang** [comment]
1.尝试跑通杜伦大学提出GANomaly：无需负例样本实现异常检测
1)对于文中使用的cifar10或者minist数据作为训练样本进行试验理解有误。以cifar10为例文中交代：Within our use of the CIFAR dataset, we again treat one class as abnormal and the rest as normal. We then detect the outlier anomalies as instances drawn from the former class by training the model on the latter labels.我在实际训练中将所有类都当做异常类来训练，试验有误，要重新试验。
2)后续想尝试使用University Baggage Anomaly Dataset(数据样式图如图2-A所示)来做个试验，但是行李箱异常数据集还没有找到相关的下载链接。
![UBA](https://user-images.githubusercontent.com/37147736/110230584-fd5e7480-7f4c-11eb-9e90-790e7e1f0fb8.PNG)
                                                                             图2-A行李箱数据集样例

**Comment by MiniBullLab** [comment]
![2021-03-08 08-59-55屏幕截图](https://user-images.githubusercontent.com/14048368/110261838-b6c05700-7fec-11eb-8f40-b3fb2a7e1984.png)
 
估计隐私数据，开源性较小。
@kingwangxiang

**Comment by MiniBullLab** [comment]
![2021-03-08 09-02-04屏幕截图](https://user-images.githubusercontent.com/14048368/110261937-0141d380-7fed-11eb-8d8c-98691c8012e0.png)

昨天提到的缺陷图像级标签数据的位置在百度网盘的上述位置。
@kingwangxiang

**Comment by lpj0822** [comment]
该问题先关闭，重新开一个任务
